File: frontend/src/api/memory.ts
import { fetchJson } from "@/api/client";

export type MemoryLayer = "" | "short" | "mid" | "long";

export interface MemoryItem {
  id: string;
  key?: string;
  value: string;
  layer?: "short" | "mid" | "long";
  count?: number;
  confidence?: number;
  importance?: number;
  mode?: string;
  memory_type?: string;
  source?: string;
  last_seen_at?: string | number;
  last_changed_at?: string | number;
  created_at?: string | number;
}

export interface MemoryAdmissionResult {
  candidate: Record<string, unknown>;
  admitted: boolean;
  reason: string;
  memory?: Record<string, unknown>;
}

export interface MemoryDecayOptions {
  short_delete_days: number;
  short_to_mid_days: number;
  short_to_mid_count: number;
  mid_to_long_days: number;
  mid_to_long_count: number;
  mid_downgrade_days: number;
  long_downgrade_days: number;
}

export async function loadMemory(limit = 240, query = "", layer: MemoryLayer = "", mode = "") {
  const params = new URLSearchParams({ limit: String(limit) });
  if (query.trim()) params.set("query", query.trim());
  if (layer) params.set("layer", layer);
  if (mode) params.set("mode", mode);
  return fetchJson<{ items: MemoryItem[]; db_path: string; mode: string }>(`/api/memory?${params.toString()}`);
}

export async function addMemory(value: string, mode = "", layer: MemoryLayer = "mid") {
  return fetchJson<{ item: MemoryItem }>("/api/memory", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ value, layer, mode }),
  });
}

export async function updateMemory(memoryId: string, patch: Partial<MemoryItem>) {
  return fetchJson<{ item: MemoryItem }>(`/api/memory/${encodeURIComponent(memoryId)}`, {
    method: "PATCH",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(patch),
  });
}

export async function deleteMemory(memoryId: string) {
  return fetchJson<{ ok: boolean }>(`/api/memory/${encodeURIComponent(memoryId)}`, { method: "DELETE" });
}

export async function runMemoryDecay(mode = "", options: Partial<MemoryDecayOptions> = {}) {
  return fetchJson<Record<string, unknown>>("/api/memory/decay", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ mode, ...options }),
  });
}

export async function testMemoryAdmission(text: string) {
  return fetchJson<{ ok: boolean; text: string; count: number; results: MemoryAdmissionResult[] }>("/api/memory/admission-test", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ text }),
  });
}

File: backend/api/memory.py
from __future__ import annotations

from fastapi import APIRouter, Body, Request

from domain.paths import MEMORY_DB
from tools.runtime_brain import admit_memory_candidate, extract_memory_candidates


def create_memory_router() -> APIRouter:
    router = APIRouter()

    @router.get("/api/memory")
    async def memory_items(request: Request, limit: int = 200, query: str = "", layer: str = "", mode: str = ""):
        return {
            "items": request.app.state.memory_store.list_memories(request.app.state.settings, limit=limit, query=query, layer=layer, mode=mode),
            "db_path": str(MEMORY_DB),
            "mode": mode or getattr(request.app.state.settings, "dialog_mode", "local"),
        }

    @router.post("/api/memory")
    async def create_memory_item(request: Request, payload: dict | None = Body(default=None)):
        payload = payload or {}
        item = request.app.state.memory_store.create_memory(payload, mode=payload.get("mode") or getattr(request.app.state.settings, "dialog_mode", "local"))
        return {"item": item}

    @router.patch("/api/memory/{memory_id}")
    async def update_memory_item(memory_id: str, request: Request, payload: dict | None = Body(default=None)):
        item = request.app.state.memory_store.update_memory(memory_id, payload or {})
        return {"item": item}

    @router.delete("/api/memory/{memory_id}")
    async def delete_memory_item(memory_id: str, request: Request):
        return {"ok": request.app.state.memory_store.delete_memory(memory_id)}

    @router.post("/api/memory/decay")
    async def decay_memory(request: Request, payload: dict | None = Body(default=None)):
        payload = payload or {}
        options = {key: value for key, value in payload.items() if key != "mode"}
        return request.app.state.memory_store.apply_decay(request.app.state.settings, mode=payload.get("mode"), options=options)

    @router.post("/api/memory/admission-test")
    async def memory_admission_test(request: Request, payload: dict | None = Body(default=None)):
        payload = payload or {}
        text = str(payload.get("text") or "").strip()
        if not text:
            return {"ok": False, "error": "text is required", "candidates": []}
        candidates = extract_memory_candidates(text)
        results = []
        for candidate in candidates:
            admitted, reason = admit_memory_candidate(candidate, text, request.app.state.settings)
            results.append({"candidate": candidate, "admitted": bool(admitted), "reason": reason, "memory": admitted})
        return {"ok": True, "text": text, "count": len(results), "results": results}

    return router
